File: trunk/ShadowFrameworkLite20/src/shadow/renderer/data/SFLibraryReferenceList.java
package shadow.renderer.data;

import java.util.LinkedList;

import shadow.system.SFInitiable;
import shadow.system.data.SFDataObject;
import shadow.system.data.SFInputStream;
import shadow.system.data.SFOutputStream;

public class SFLibraryReferenceList<T extends SFInitiable> extends LinkedList<SFLibraryReference<T>> implements SFDataObject{

	private static final long serialVersionUID=0;
	
	private SFLibraryReference<T> type;
	
	public SFLibraryReferenceList(SFLibraryReference<T> type) {
		super();
		this.type = type;
	}
	
	public void add(String reference){
		SFLibraryReference<T> libraryReference = type.clone();
		libraryReference.setReference(reference);
		add(libraryReference);
	}

	@Override
	public void readFromStream(SFInputStream stream) {
		int n=stream.readShort();
		for (int i = 0; i < n; i++) {
			SFLibraryReference<T> reference=type.clone();
			add(reference);
			reference.readFromStream(stream);
		}
	}
	
	@Override
	public void writeOnStream(SFOutputStream stream) {
		stream.writeShort((short)(size()));
		for (int i = 0; i < size(); i++) {
			get(i).writeOnStream(stream);
		}
	}
	
	@Override
	public SFLibraryReferenceList<T> clone(){
		return new SFLibraryReferenceList<T>(type);
	}
	
}

File: trunk/ShadowFrameworkJS/src/shadow/renderer/data/SFLibraryReferenceList.js


function SFLibraryReferenceList(type){
	this.type=type;
	this.list=new Array();
}


SFLibraryReferenceList.prototype["add"]=function(reference){
	//var libraryReference = this.type.clone();
	//libraryReference.setReference(reference);
	this.list.push(reference);
};

SFLibraryReferenceList.prototype["readFromStream"]=function(stream){
	var n=stream.readShort();
	for (var i = 0; i < n; i++) {
		var reference=this.type.clone();
		this.add(reference);
		reference.readFromStream(stream);
	}
};

SFLibraryReferenceList.prototype["clone"]=function(){
	return new SFLibraryReferenceList(this.type);
};
	

SFLibraryReferenceList.prototype["get"]=function(index){
	return this.list[index];
};

SFLibraryReferenceList.prototype["size"]=function(){
	return this.list.length;
};